### PR TITLE
[NFC][SYCL] Switch to std:: equivalents for utilities in stl_type_traits.hpp

### PR DESCRIPTION
**Pull request by aelovikov-intel**
Re-commits #7628 and #7668 that were previously reverted due to build issues on RHEL systems. The issues were fixed by introducing `SYCL_LIBDEVICE_GCC_TOOLCHAIN` cmake configuration option in #7771.

### CODE REVIEW

File: sycl/include/sycl/ext/intel/esimd/detail/memory_intrin.hpp [hunk 1]
```diff
@@ -70,7 +70,7 @@ constexpr unsigned int ElemsPerAddrEncoding() {
   else if constexpr (ElemsPerAddr == 4)
     return 2;
 
-  // other cases not needed since enable_if disallows other values
+  // other cases not needed since std::enable_if disallows other values
```
Comment (sarnex) [inline sycl/include/sycl/ext/intel/esimd/detail/memory_intrin.hpp:73]: prolly revert this?

**Review by sarnex — APPROVED** [review]
esimd lgtm, only sed fallout :)

**Comment by aelovikov-intel** [comment]
Thanks for the quick review! I noticed those as well, but I think it makes sense to keep the changes.

**Comment by sarnex** [comment]
ok sounds good, no strong opinion

**Comment by aelovikov-intel** [comment]
I know that review has already started, but I think rebase for this kind of PR is slightly better than merge.

**Review by bso-intel — APPROVED** [review]

**Review by v-klochkov — APPROVED** [review]

**Comment by aelovikov-intel** [comment]
CUDA failure on `SYCL :: Regression/same_unnamed_kernels.cpp` is known - https://github.com/intel/llvm/issues/9174.

@intel/llvm-gatekeepers , this PR is ready.

**Pull request merged**